features/202-session-management/design.md

```diff
@@ -0,0 +1,163 @@
+mutation.
+
+Gateway is responsible to update last access time, last IP, and last user agent
+of session after processing the request, regardless whether the request is
+successful.
+
+
+## Session Management
+
+APIs would be provided to manage sessions of a user.
+
+Session list API returns a list of valid session of a user. Each list entry
+contains the session ID and session attributes. Access token / refresh token of
+the session would not be included.
+
+Revoke sessions API revokes a specific session, or all sessions other than the
+requesting session. A `session_delete` event would be generated for all sessions
+to be revoked.
+
+
+## Access / Refresh Token
+
+Each session would have an access token, and optionally a refresh token. Both
+type of tokens are opaque: token formats are unspecified and developer should
+not attempt to interpret the content of tokens.
+
+Access token would always have a lifetime; access token would be treated as
+invalid after its expiry.
+
+Refresh token can be used to obtain a new access token. If the old access token
+is still valid, the old access token is invalidated; there would be at most one
+valid access token for a session at any time.
+
+Refresh token would always have a maximum lifetime.
+
+Optionally, a session idle timeout can be specified: refresh token (or access
+token if refresh token is disabled) must be used at least once before the
+timeout, otherwise the session would be expired.
+
+A session is invalidated if its identity is deleted, or its user is disabled.
+A session is expired if its refresh token is expired, or its access token is
+expired (if refresh token is disabled).
+If a session is invalidated, expired, revoked, or logout, its associated access
+token / refresh token would be treated as invalid.
+
+Client-side SDK should handle refreshing of expired access token automatically
+if refresh token is available; SDKs would not expose concept of access token /
+refresh token; however, developers should still handle situation where session
+is no longer valid due to various reasons.
+
+
+## Session Cookie
+
+Optionally, developer can choose to use cookies to transport session tokens, in
+order to facilitate server-side rendering.
+
+If developer choose to enable session cookie:
+- Refresh token would not be issued. (See [Appendix](#session-cookie-and-refresh-token))
+- Access token would be transported in session cookie.
+- Access token would not be returned from APIs.
+- Gateway would read access token from cookie only, ignoring `Authorization`
+  header.
+
+### Security Consideration
+
+- HTTPS is required.
+- Session cookie is HTTP-only; web app and web SDK would not have access to the
+  session tokens.
+- Session cookie is set with `SameSite=lax`; generally, cross-domain requests
+  would not contain the cookie.
+- Session cookie would not be shared across custom domains.
+- App developers are responsible to prevent CSRF in their services.
```

Comment:
I haven't think it through, but if people would submit http req to Skygear's Cloud Functions or Microservices end point directly, maybe it would be great if we can have basic CRSF prevention support (e.g pre-generated CSRF token in cookies, and validate it on gateway)

We don't need it in version 1, but good to have it in spec and open issue first?

Reply:
I think supporting CSRF prevention would be out of scope for this spec. We can open an issue for it for now.

Reply:
Depends on if this spec is about the session management API or session itself, if it is the session itself (it seems yes, the spec got access token and refresh token together), I think CSRF is within the spec as it is closely related.

But if that's too much works, please open a feature issue for it, as we are constrained by time anyway.

Reply:
opened #351